### PR TITLE
FASTA type

### PR DESCRIPTION
**Pull request by bicycle1885**
This adds type-specific parser for FASTA formats and fixes the sequence type inference reported in #178.

I've found that the naive Bayes classifier is too unreliable to classify sequence types and introduced a simple heuristic to improve the accuracy (thank you @timbitz!):
```julia
    if ACGTUN ratio > 0.95
        if #T ≥ #U
            return DNASequence
        else
            return RNASequence
        end
    else
        return AminoAcidSequence
    end
```

However, this heuristic is still not perfect when a sequence is too short or too skewed. So, I made it possible to specify the sequence type when creating a FASTA parser:
```julia
# set a sequence type as the third argument (optional)
for record in open("some.fasta", FASTA, AminoAcidSequence)
   # record.seq is surely AminoAcidSequence
end
```
The advantage of it is that Julia can infer the sequence type and will generate a more optimized program due to the type stability.

### CODE REVIEW

**Comment by vchuravy** [comment]
I think that being able to specify the sequence type is really good. The automatic inference is cool, but if I am working on a pipeline I know what data I move through the system and I would rather have the system loudly fail, than silently move on and create invalid results.

**Comment by bicycle1885** [comment]
@vchuravy Thanks :).

This is now ready to review. I'd like to merge this if you like the API.

File: src/Ragel.jl [hunk 1]
```diff
@@ -229,18 +229,18 @@ end
...
 # ----------------------------------
 
-function Base.open{T<:FileFormat}(filename::AbstractString, ::Type{T}; args...)
+function Base.open{T<:FileFormat}(filename::AbstractString, ::Type{T}, args...; kwargs...)
     memory_map = false
     i = 0
-    for arg in args
+    for arg in kwargs
         i += 1
         if arg[1] == :memory_map
```
Comment (Ward9250) [inline src/Ragel.jl]: Is it mandatory that memory_map be the first two kwargs? If so can this be changed? I can imagine users occasionally getting the order wrong.

**Comment by Ward9250** [comment]
LGTM :+1:

File: src/Ragel.jl (hunk 1 above)
Comment (bicycle1885) [inline src/Ragel.jl]: This does not mean `memory_map` must be the first argument. `arg[1]` is a key of a keyword argument and `arg[2]` will be a value of it. This loop tries to find a pair of `memory_map => true/false` and thus `memory_map` can be located at any position.

Comment (Ward9250) [inline src/Ragel.jl]: Ah I see each iteration `arg` looks like a tuple. I follow you now. Seems mornings are not good to me reading code.

**Comment by timbitz** [comment]
Sweet.

**Pull request merged**